_drafts/device-logging.md

```diff
@@ -0,0 +1,295 @@
+tags: [android, debugging, logging]
+---
+
+Logging on embedded devices comes with a host of interesting, unique challenges that aren't found on the server &mdash; devices are generally more memory, bandwidth, space, and CPU constrained than a server, and devices often take weeks, months, or years to update, rather than days or hours. This has interesting ramifications on how we log: we can't log as much data as we'd want to, logs have a more finite lifespan, and the format of logs are much more permanent, so if we don't log the right data in the right way to begin with, it can be months before we ship the next release with new logs. All that being said, there is still a lot of literature and learnings from logging on the server that we can also apply on the client side.
+
+<!-- excerpt start -->
+
+In this article, I discuss some of the problems I’ve encountered in both producing and analyzing device logs over the years, on single digit number of devices up to hundred of thousands, and concepts I’ve applied to manage that complexity. My background is in Android development, so some of the examples may skew towards much more powerful hardware than simple microcontrollers.
+
+<!-- excerpt end -->
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+---
+
+## Log Collection
+
+As is common in embedded development, there are many interesting decisions to be made long before any code is actually written, or any logs actually logged. Before we can use the logs, we first need to design how the logs are stored on the device, and then how to get access to them.
+
+### Log Cache Size
+
+One of the fundamental constraints to work around with device logging is how much disk space to allocate just for logging. Unlike on servers where the log cache is basically infinite (and you can often pay more for the storage), most logging systems on hardware will reserve a fixed amount of disk space or memory in the form of a rotating buffer of logs. When you log past the limit, the logs in the oldest buffer gets written over.
+
+The amount of disk space chosen has a large influence on logging &mdash; it affects how verbose your logs can be, how long your logs are be retained, what information you'll want prioritize logging, how you'll store the logs, and more. You’ll want to choose the maximum amount of space you can afford, where the logs you get back contain enough useful information at the timeframe of the incident. But you'll also want to change how you log to accomodate for the amount of space, such as using an abbreviated short form instead of full words.
+
+I've been lucky in that embedded Android devices are generally not very disk space constrained, so I’ve had the benefit of having 10's of MBs of logs to store (and comb) through. But even so, that would only constitute maybe 2-3 hours of logs, depending on how busy the device was, and there was a constant trade-off with how much information was being logged.
+
+### On-location v. Remote Collection
```

Comment:
TIL `v.` is used in legal docs to shorten versus... and I don't like it 🙃. This would be personal preference I prefer `vs.`.

Reply:
Fair point, now that you've mentioned it, I can't unsee the legal usage. `vs.` works for me